### PR TITLE
[#17] [iOS] [UI] As a user, I can see Survey selection page

### LINKED ISSUE #17: [iOS] [UI] As a user, I can see Survey selection page
## Why

User should be able to select a survey from a list of surveys provided by the API.

## Acceptance Criteria

- Show Survey large image.
- Show Survey title.
- Show survey description.
- Show page indicator.
- Show `>` button.

Screen Shot 2565-09-12 at 16 34 12

## Resources
[Figma](https://www.figma.com/file/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App?node-id=30%3A46)

### PR DESCRIPTION
**Pull request by blyscuit**
close #17 

## What happened

Add survey selection page.

## Insight

- Create `FadePaginationView` for full screen pagination with fade animation.

## Proof of Work

https://user-images.githubusercontent.com/6356137/192745607-088825a3-8cc8-4c21-9541-cbd929bd1bf4.mov

https://user-images.githubusercontent.com/6356137/192745691-3caffb2b-848f-4d1c-91b7-f091453f7ff3.mov

### CODE REVIEW

File: iosApp/Survey/Sources/Presentation/Modules/Login/LoginView.swift [hunk 1]
```diff
@@ -8,6 +8,11 @@
 
 import SwiftUI
 
+protocol LoginCoordinator {
+
+    func showSurvey()
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Modules/Login/LoginView.swift]: I think you should consider to name from `surveySelection` or `survey` here  to `Home` as that would be a more common flow that we go from Login screen to Home screen? What do you think? 🤔
```suggestion
    func showHome()
```

File: iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import SwiftUI
+
+struct SurveyItemView: View {
+
+    let survey: SurveyUI
+
+    var body: some View {
+        ZStack {
+            GeometryReader { geo in
+                KFImage(URL(string: survey.largeImageUrl) ?? URL(fileURLWithPath: ""))
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift]: May we we can turn this part - `URL(string: survey.largeImageUrl) ?? URL(fileURLWithPath: "")` into an extension as there are chances we will need to reuse for intializing the `URL` 🤔

File: iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                    .font(.regularBody)
+                    .lineLimit(2)
+                    .frame(maxWidth: .infinity, alignment: .leading)
+                    .offset(y: descriptionOffset)
+                    .accessibility(.surveySelection(.detailText))
+            }
+            .id(survey.id)
+            .padding(.leading, 20.0)
+            .padding(.trailing, 76.0)
+            .accessibilityElement(children: .contain)
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift]: Just curious, why do we need this `.accessibilityElement(children: .contain)`? Is it for testing? 🤔

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import SwiftUI
+
+struct FadePaginationView<T>: View {
+
+    @Binding private var currentPage: Int
+    @State private var currentVisible: Double = 1.0
+    @State private var nextPage = 0
+    private var turnLength: Double = 160.0
+    private var turnSpeed: Double = 700.0
+    @GestureVelocity private var velocity: CGVector
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Can we have some in between line breaks for these variables for clean separation per type?

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        self.items = items
+    }
+
+    func forwardPage() {
+        currentPage = max(0, currentPage - 1)
+    }
+
+    func previousPage() {
+        currentPage = min(items.count - 1, currentPage + 1)
+    }
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: These functions can be `private`, right?

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                .gesture(
+                    DragGesture(minimumDistance: 6.0, coordinateSpace: .local)
+                        .onChanged { value in
+                            switch value.translation.width {
+                            case ...0: nextPage = min(items.count - 1, currentPage + 1)
+                            case 0...: nextPage = max(0, currentPage - 1)
+                            default: return
+                            }
+                            let percentage = turnLength * 1.17 / abs(value.translation.width)
+                            currentVisible = 1.0 * max(0.0, percentage)
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: I think the logic here for calculating the `percentage` and `currentVisible` is a bit ambiguous such as: 
- What does the value `1.17` mean? 
- Why is the `minimumDistance` equaling to `6.0`? 
- What is `percentage` here for (percentage of the screen width or something)? 
- etc.

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+//  Created by Bliss on 28/9/22.
+//  Copyright © 2022 Nimble. All rights reserved.
+//
+
+import SwiftUI
+
+struct FadePaginationView<T>: View {
+
+    @Binding private var currentPage: Int
+    @State private var currentVisible: Double = 1.0
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: `currentVisible` sounds a bit weird as there is not a noun presenting the variable. Please correct me if I am wrong but it should be `currentVisibleThreshold` or `currentVisibility`, right?

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    @State private var nextPage = 0
+    private var turnLength: Double = 160.0
+    private var turnSpeed: Double = 700.0
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Naming suggestions 🙏
```suggestion
    private var turningLength: Double = 160.0
    private var turningSpeed: Double = 700.0
```

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                                    forwardPage()
+                                } else {
+                                    switch value.translation.width {
+                                    case ...(-turnLength):
+                                        previousPage()
+                                    case turnLength...:
+                                        forwardPage()
+                                    default: break
+                                    }
+                                }
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Can we have some comments to explain each `if else` case here for more clarity? 🙏

File: iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift (hunk 3 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift]: This one is unnecessary, I will remove it. `accessibilityElement(children` is how SwiftUI assign accessibility id to children of a view, without `.accessibilityElement(children: .contain)` all the children views will have the same id as the parent.

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift (hunk 9 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Updated in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/de04d90f5ac9e5bbc3b76ee7f3127b8ae887468e

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift (hunk 8 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Updated in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/de04d90f5ac9e5bbc3b76ee7f3127b8ae887468e

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift (hunk 5 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Updated in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/de04d90f5ac9e5bbc3b76ee7f3127b8ae887468e

File: iosApp/Survey/Sources/Presentation/Modules/Login/LoginView.swift (hunk 1 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Modules/Login/LoginView.swift]: Updated in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/de04d90f5ac9e5bbc3b76ee7f3127b8ae887468e

File: iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift (hunk 2 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Modules/SurveySelection/SurveyItemView.swift]: Updated in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/de04d90f5ac9e5bbc3b76ee7f3127b8ae887468e

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift (hunk 6 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Updated in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/de04d90f5ac9e5bbc3b76ee7f3127b8ae887468e

File: iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift (hunk 7 above)
Comment (blyscuit) [reply iosApp/Survey/Sources/Presentation/Views/FadePaginationView.swift]: Updated in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/de04d90f5ac9e5bbc3b76ee7f3127b8ae887468e

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/presentation/model/SurveyUI.kt [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package co.nimblehq.blisskmmic.presentation.model
+
+data class SurveyUI (
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/presentation/model/SurveyUI.kt]: On Android, we have `SurveyUiModel`, so you can follow it

Comment (minhnimble) [reply shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/presentation/model/SurveyUI.kt]: @suho and maybe we can apply that for iOS also, just a thought 😆

**Review by minhnimble — APPROVED** [review]
Apart from @suho comment, rest LGTM 🚀

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/presentation/model/SurveyUI.kt (hunk 10 above)
Comment (blyscuit) [reply shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/presentation/model/SurveyUI.kt]: Renamed in https://github.com/blyscuit/kmm-swiftUI-IC/pull/57/commits/e9922606fe89a34ab57dcee3942dd7ded99ca3a6

**Review by edgarss — APPROVED** [review]
LGTM

**Pull request merged**